Spi map example

landmanbester opened the pull request:
Let's get this into master so I can get some feedback. It seems to be working on simulated data at least

sjperkins commented on africanus/model/spi/examples/simple_spi_fitter.py:
```diff
@@ -0,0 +1,402 @@
+    return p
+
+
+args = create_parser().parse_args()
```
I think it would be safer if the contents of the script were placed in a function lilke so:

```python
def main():
   ....

if __name__ == "__main__":
    main()
```

Briefly, this allows the module to be imported without running the script. https://stackoverflow.com/a/19578357/1611416, which can be useful for tools which inspect python code (such as pytest, probably)

sjperkins commented on africanus/model/spi/examples/README.rst:
```diff
@@ -0,0 +1,33 @@
+
+Fits a simple spectral index model to image cubes. Usage is as follows
+
+$ ./simple_spi_fitter.py --fitsmodel=/path/to/model.fits 
```
This is a  nice description of the example. You could introduce some restructured text code formatting here. Either:

```
Fits a simple spectral index model to image cubes. Usage is as follows::

    $ ./simple_spi_fitter.py --fitsmodel=/path/to/model.fits 

```

or 

```
Fits a simple spectral index model to image cubes. Usage is as follows

.. code-block:: bash

    $ ./simple_spi_fitter.py --fitsmodel=/path/to/model.fits 

```
should do it

landmanbester commented:
Ok, I didn't like the interface to the pyfftw objects and there was a significant overhead to using the pyfftw.interfaces functions so I switched to using pypocketfft. I don't think pypocketfft is on PYPI yet though, you need to install it manually from the repo. @sjperkins do you think we should make this an optional requirement? We will also need it for the gridder and I would like to use it for the multi-scale algorithm I am putting in.
I've also implemented your changes and marked them as resolved but for some reason the tests are not getting triggered on the latest changes. Not sure why. Any ideas?

sjperkins commented:
> Ok, I didn't like the interface to the pyfftw objects and there was a significant overhead to using the pyfftw.interfaces functions so I switched to using pypocketfft. I don't think pypocketfft is on PYPI yet though, you need to install it manually from the repo. @sjperkins do you think we should make this an optional requirement? We will also need it for the gridder and I would like to use it for the multi-scale algorithm I am putting in.

Are there any codex API's that need pocketFFTW, at present or just the example? If it's just the example, then I don't think it needs to be an optional requirement in the setup.py yet. Rather add it in a gridder/multi-scale PR for e.g.

> I've also implemented your changes and marked them as resolved but for some reason the tests are not getting triggered on the latest changes. Not sure why. Any ideas?

Ah I think this is because I disabled [build on PR](https://docs.travis-ci.com/user/web-ui/#build-pushed-pull-requests) no travis because I was trying to avoid [double builds](https://docs.travis-ci.com/user/pull-requests/#double-builds-on-pull-requests). I've re-enabled this on travis now. I think you may need to push another commit up to trigger this though?

sjperkins commented:
> Are there any codex API's that need pocketFFTW, at present or just the example? If it's just the example, then I don't think it needs to be an optional requirement in the setup.py yet. Rather add it in a gridder/multi-scale PR for e.g.

Apologies I should have read what you said more closely. We can't put it in setup.py because its not on pypi. I think you should do something like:

```python
try:
   import pocketfft
except ImportError as e:
    opt_import_error = ImportError("Unable to import pocketfft.
                       "Please manually install it from http://pocket.fftw.url/")
else:
    opt_import_error = None


@requires_optional("pocketfft", opt_import_error)
def function(*args, **kwargs):
    ....
```

landmanbester commented:
@sjperkins looks like we broke test_requirements. Can you have a look please?

sjperkins commented:
The `len(import_errors) == 0` case in ``_missing_packages`` needs to be catered for, otherwise it returns None.

landmanbester commented:
Ah yes, of course. I think it should pass now but note that the function now returns a list of packages instead of a tuple. Will that be problem?

sjperkins commented on africanus/util/requirements.py:
```diff
@@ -15,6 +15,9 @@ def _missing_packages(fn, packages, import_errors):
                 "%s" % (fn, packages, import_err_str))
+    else:
+        return ("%s requires installation of the following packages: %s. "
+                % (fn, *packages))
```
> Ah yes, of course. I think it should pass now but note that the function now returns a list of packages instead of a tuple. Will that be problem?

Do you mean passes a list instead of a tuple?.

I think this should work:

```python
     return ("%s requires installation of the following packages: %s. "
                % (fn, tuple(packages)))
```

landmanbester replied:
Yep. I just changed it to unpack the list and modified the string it compares to accordingly.
